topics/products/phpstorm/php_open_api_php_type_providers.md

```diff
@@ -0,0 +1,303 @@
+definitely say that a numeric literal is of type `int`.
+
+```php
+<?php
+
+function foo(): string { return "Hello World!"; }
+```
+
+Here, since the `foo()` function has a type hint, the return type is also Complete.
+
+### Incomplete types
+
+Incomplete types are types that require additional resolving and contain additional information for accurate type inference.
+
+Suppose we have the following code:
+
+```php
+<?php
+class Foo {
+  public function f() {}
+}
+
+function foo(): Foo { return new Foo(); }
+
+foo()->f();
+```
+
+Here PhpStorm must somehow express the type of the expression `foo()` during indexing, however,
+during indexing, PhpStorm cannot refer to the definition of the `foo()` function to return a specific type.
+
+So PhpStorm creates an Incomplete type into which it writes all the information it might need to resolve the type.
+In this case, the name of the function and its arguments, so PhpStorm will create a new Incomplete type `#琁\foo.`.
+
+#### Incomplete Types Structure
+
+At the very beginning there is a `#` character, which is a marker that the type is Incomplete.
+
+It is followed by the type provider's unique key.
+According to it, you can understand that this type was created by the provider of the desired type.
+
+The rest of the line is the encoded information.
+In the example above, it's the fully qualified name and a dot.
+
+The created Incomplete type will be processed further when the indexing process is over.
+
+During indexing, the IDE collects information about all types of elements in this way and stores them in the index.
+When there is a need for the type of some expression, the type obtained at the Incomplete indexing stage is passed for resolving.
+
+#### Incomplete Types Resolving
+
+Incomplete types are resolved if it was necessary to infer the element type,
+since indexing has already been completed at this point, PhpStorm have access to all information,
+for example, functions and their return types.
+
+Receiving Incomplete type PhpStorm gets from it all the information that was placed there during the indexing stage.
+In the example above this is the name of the function and its arguments.
+
+With this information, PhpStorm can access the index and get the information it needs.
+In the example above, this is the return type of the function.
+
+As a result, the resulting Complete type is returned and will be used by PhpStorm for auto-completion, checks, etc.
+
+### Union Types
+
+Since PHP is a dynamically typed language, at the type inference stage PhpStorm can get a situation
+where the type can be either one or the other.
+For example:
+
+```php
+<?php
+
+$a = 100;
+if (rand(0, 10) > 5) {
+  $a = "Hello World!".
+}
+
+$a; // (1)
+```
+
+In (1), the variable `$a` will be of type `int|string` because PhpStorm can't deduce exactly which branch the execution will take.
+
+PhpStorm stores these types by separating each of the individual types with a `|` character.
```

Comment:
It's not correct, `|` used only for representation purposes. Multiple types are stored as multiple strings inside `PhpType`

Reply:
Fixed.